### PR TITLE
disable db delete when response to notify 408/481 for already expired subscriptions

### PR DESCRIPTION
**Pull request by aniketkuwarvonage**

### CODE REVIEW

File: modules/presence/hash.c [hunk 1]
```diff
@@ -793,4 +793,48 @@ int delete_cluster_query(str* pres_uri, int event, unsigned int hash_code)
...
+	{
+		if(s->to_tag.len== to_tag.len &&
+				strncmp(s->to_tag.s, to_tag.s, to_tag.len)== 0)
+		{	
+			found = 1;
+			if(s->expires == 0 && 
+					s->status == TERMINATED_STATUS &&
+						strcasecmp(s->reason.s, "timeout") == 0)
+			{
+				LM_WARN("sub found and expired to_tag [%.*s]\n",to_tag.len, to_tag.s);
```
Comment (WeipingGuo) [inline modules/presence/hash.c:820]: Those messages are good for debugging purpose, is it too much for live system?

Comment (tommybrecher) [reply modules/presence/hash.c]: Maybe `LM_INFO` is a good logging level for those messages?

We want to be able to check if we are hitting this condition but don't the overhead of `LM_DEBUG`

Comment (aniketkuwarvonage) [reply modules/presence/hash.c]: Changed log level to INFO

**Review by WeipingGuo — APPROVED** [review]

File: modules/presence/hash.c [hunk 2]
```diff
@@ -793,4 +793,48 @@ int delete_cluster_query(str* pres_uri, int event, unsigned int hash_code)
...
 }
 
+/* 
+ * entry must be locked before calling this function 
+ * Returns 1 if 
+            - entry exisits subscription is expired. 
+			- entry does not exist.
+ * Returns 0 if entry is found and subscription is active. 
+*/
+int is_sub_expired(shtable_t htable,unsigned int hash_code,str to_tag)
```
Comment (tommybrecher) [inline modules/presence/hash.c:803]: Not urgent but would be nice to format code additions (space between names, space before and after operators, etc) to match the style-guide of upstream

Provided a few examples but non-blocking
```suggestion
int is_sub_expired(shtable_t htable, unsigned int hash_code, str to_tag)
```

File: modules/presence/hash.c [hunk 3]
```diff
@@ -793,4 +793,48 @@ int delete_cluster_query(str* pres_uri, int event, unsigned int hash_code)
...
+{
+	subs_t* s= NULL;
+	int is_expired = 0, found = 0;
+
+	s = htable[hash_code].entries->next;
+
+	while(s)
+	{
+		if(s->to_tag.len== to_tag.len &&
```
Comment (tommybrecher) [inline modules/presence/hash.c:812]:
```suggestion
		if(s->to_tag.len == to_tag.len &&
```

File: modules/presence/hash.c [hunk 4]
```diff
@@ -793,4 +793,48 @@ int delete_cluster_query(str* pres_uri, int event, unsigned int hash_code)
...
+				strncmp(s->to_tag.s, to_tag.s, to_tag.len)== 0)
+		{	
+			found = 1;
+			if(s->expires == 0 && 
+					s->status == TERMINATED_STATUS &&
+						strcasecmp(s->reason.s, "timeout") == 0)
```
Comment (tommybrecher) [inline modules/presence/hash.c:818]: Line continuation (stay indented by 2 tabs)
```c
			if(s->expires == 0 && 
			        s->status == TERMINATED_STATUS && 
			        strcasecmp(s->reason.s, "timeout") == 0)
```

**Review by tommybrecher — APPROVED** [review]
Looks good to me, good stuff.

I do have some cosmetic suggestions if you feel like formatting a little to match upstream style guide but certainly not blocking

**Review by kxin-vonage — APPROVED** [review]
Good stuff!

**Comment by WeipingGuo** [comment]
@kxin-vonage @aniketkuwarvonage  We need this merged into master_2.4.9 before 2.45 AMI rebuild

**Pull request merged**